### ISSUE TITLE
Training Loss problem

### ISSUE THREAD

**Issue by RichardYang88** [hyunwoongko/transformer#33]
1. Training Loss Discrepancy
During training, I observed that the test set loss is consistently lower than the training set loss, which contradicts the loss-epoch curves shown in your paper/repository. Could this indicate a potential issue in the code (e.g., data leakage, incorrect evaluation setup) or a submission error? I followed the training protocol as described, but the anomaly persists.

2. Code-Specific Issues
Issue 1: Data Type Error in make_trg_mask Function
In transformer.py, the make_trg_mask function raises a data type error. The mask should be of torch.bool type. I resolved it by modifying the line:

# Original (error-prone):  
trg_sub_mask = torch.tril(torch.ones(trg_len, trg_len)).type(torch.ByteTensor).to(self.device)  

# Fixed version:  
trg_sub_mask = torch.tril(torch.ones(trg_len, trg_len)).unsqueeze(0).unsqueeze(0).type(torch.bool).to(self.device) 
 
Issue 2: Dataset Download Failure
The dataset failed to download automatically via the provided script. To proceed, I manually downloaded and extracted the data from the original source:
https://github.com/multi30k/dataset.git  
You may want to update the dataset fetching logic to avoid this issue for future users.

Could you kindly verify if these adjustments align with your intended implementation? Any insights or suggestions would be greatly appreciated.

Thank you for your time and support!

**Comment by RichardYang88** [comment]
The training loss is as follows. The training loss is lower than the test loss. Obviously, something went wrong in the source code.

Epoch: 1 | Time: 0m 57s
        Train Loss: 6.762 | Train PPL: 864.715
        Val Loss: 5.940 |  Val PPL: 380.029
        BLEU Score: 0.000

Epoch: 2 | Time: 0m 56s
        Train Loss: 5.912 | Train PPL: 369.564
        Val Loss: 6.014 |  Val PPL: 409.010
        BLEU Score: 4.482

Epoch: 3 | Time: 0m 56s
        Train Loss: 5.551 | Train PPL: 257.571
        Val Loss: 5.312 |  Val PPL: 202.836
        BLEU Score: 11.047

**Comment by Deaddawn** [comment]
I know what's going on here. The final output of the transformer does not use softmax for normalization in his code. The values directly output by the linear layer used fro training are definitely incorrect. @RichardYang88
